### PR TITLE
fix(semantic_analyzers): fix the false positive for `noConstAssign`

### PR DESCRIPTION
**Pull request by mzbac**
## Summary

This PR is for fixing https://github.com/rome/tools/issues/3728. The current implementation of noConstAssign did not skip function assignment that caused a false error for the lint rule.

## Test Plan
I added two new test cases to cover this edge case.

### CODE REVIEW

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 1]
```diff
@@ -61,13 +61,14 @@ impl Rule for NoConstAssign {
...
         let declared_binding = model.declaration(node)?;
-        if let Some(variable_declaration) = declared_binding
-            .syntax()
-            .ancestors()
-            .find_map(|ancestor| JsVariableDeclaration::cast_ref(&ancestor))
-        {
-            if variable_declaration.is_const() {
-                return Some(declared_binding.syntax().text_trimmed_range());
+        for node in declared_binding.syntax().ancestors() {
+            if let Some(_) = JsFormalParameter::cast_ref(&node) {
```
Comment (ematipico) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:65]: You can use 

```rust
JsFormalParameter::can_cast(&node)
```

Which returns a `bool`. It seems that's what you need.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 2]
```diff
@@ -61,13 +61,14 @@ impl Rule for NoConstAssign {
...
 
         let declared_binding = model.declaration(node)?;
-        if let Some(variable_declaration) = declared_binding
-            .syntax()
-            .ancestors()
-            .find_map(|ancestor| JsVariableDeclaration::cast_ref(&ancestor))
-        {
-            if variable_declaration.is_const() {
-                return Some(declared_binding.syntax().text_trimmed_range());
+        for node in declared_binding.syntax().ancestors() {
```
Comment (leops) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:64]: I think this is only solving a small part of the problem. The underlying issue is that visiting all the ancestor nodes until a `JsVariableDeclaration` is found will cause any constant declaration to "bleed" onto all the identifier bindings that are part of the initializer expression, not just `JsFormalParameter`. For instance all these case will also raise a similar false positive:

```js
const JsVariableDeclaration1 = function JsFunctionExpression() {
    JsFunctionExpression = 1;
};

const JsVariableDeclaration2 = class JsClassExpression {};
JsClassExpression = 2;

const JsVariableDeclaration3 = () => {
    try {
        func();
    } catch(err) {
        err = 3;
    }
}

const JsVariableDeclaration4 = (...rest) => {
	rest = 4;
};
```

Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:64]: Thank you for the feedback. It looks like we will have to do the ancestor search in block scope. Let me see if I can do it easily.

Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:68]: @leops 

I only figured out a hardcoded approach to address those cases. Please let me know what you thought.

Comment (leops) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs]: From reading the code it seems like this approach wouldn't work for bindings that are nested in assignment patterns, leading to false negatives in case like `const [a, {  b }] = expr; a = expr; b = expr;`
But in general, due to the complexity of correctly handling all the corner cases of JavaScript syntax being able to look up the `JsAnyDeclaration` node for a given binding is probably something that should be provided by the semantic model so we don't have to reimplement it on all the rules that need something like this (`useConst` for instance does the inverse operation and needs to enumerate all the bindings created by a variable declaration node). @xunilrj Do you have an idea on how this could work ?

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 3]
```diff
@@ -65,19 +62,20 @@ impl Rule for NoConstAssign {
...
-                || JsFunctionExpression::can_cast(node.kind())
-            {
-                return None;
-            }
-            if let Some(variable_declaration) = JsVariableDeclaration::cast_ref(&node) {
-                if variable_declaration.is_const() {
-                    return Some(declared_binding.syntax().text_trimmed_range());
+        let parent = declared_binding.syntax().parent()?;
+
+        match parent.kind() {
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:67]: Does this work with destructuring? Would you mind adding a test for it

```
const { a, { b } } = { ... }
```

Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:67]: will add it, thanks

Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs]: Hi @MichaReiser,
I have tried your suggestion, however the JsVariableDeclaration and JsForVariableDeclaration would be one to five levels up from the non-JsAnyBindingPattern parent. I feel using the hardcoded position may not be the best approach. So I only change my solution upwards when  non-JsAnyBindingPattern parent is JsVariableDeclarator/JsArrayBindingPatternElementList/JsObjectBindingPatternShorthandProperty.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 4]
```diff
@@ -61,13 +65,28 @@ impl Rule for NoConstAssign {
         let model = ctx.model();
 
         let declared_binding = model.declaration(node)?;
-        if let Some(variable_declaration) = declared_binding
+
+        let js_any_binding_parent = declared_binding
             .syntax()
             .ancestors()
-            .find_map(|ancestor| JsVariableDeclaration::cast_ref(&ancestor))
+            .find(|n| !JsAnyBindingPattern::can_cast(n.kind()))?;
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:72]: It should be possible to merge this iteration with the one coming after. 

The goal of the first navigation is to skip any binding parents:

```
let non_binding = declared_binding.syntax().ancestors().skip_while(|node| 
	node.kind().is_list() 
		|| JsAnyBindingPattern:;can_cast(node.kind()) 
		|| JsAnyArrayBindingPatternElement::can_cast(node.kind())
		|| JsAnyObjectBindingPatternMember::can_cast(node.kin())
```

From there, it should be possible to go one level up to retrieve the variable declaration

Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs]: @MichaReiser 

I updated the PR as you suggested, but there is an additional one level up look up for  the JsVariableDeclaratorList.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 5]
```diff
@@ -61,13 +64,39 @@ impl Rule for NoConstAssign {
...
-            .find_map(|ancestor| JsVariableDeclaration::cast_ref(&ancestor))
-        {
-            if variable_declaration.is_const() {
-                return Some(declared_binding.syntax().text_trimmed_range());
+
+        if let Some(possible_declarator) = declared_binding.syntax().ancestors().find(|node| {
+            !matches!(
+                node.kind(),
+                JS_OBJECT_BINDING_PATTERN_SHORTHAND_PROPERTY
+                    | JS_OBJECT_BINDING_PATTERN_PROPERTY_LIST
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:72]: Using the union types (`AnyObjectBindingPatternMember` and  `AnyArrayBindingPatternMember`) instead of listing all kinds explicitly is more future-proof and ensures that the code works for rest elements: `const { a, ...rest } = c`

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs (hunk 2 above)
Comment (xunilrj) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs]: Today the semanic model has a ill named `declaration()` that returns the associated binding of a reference. From the binding to the actual declaration is a syntatic navigation.

What I am thinking is:

- rename this method to `binding()` or someting like that;
- extension method on JsIdentifierBinding (and similars) to return JsAnyDeclaration;
- semantic model can very easily return a list of all bindings and give easy access if they have writes or not.

We can use these to create this rule:

- Query the Semantic Model;
- Iter all bindings, filter those with at least one write;
- Navigate to declaration and test if is const or not.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 6]
```diff
@@ -61,13 +64,39 @@ impl Rule for NoConstAssign {
...
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
+                if let Some(js_for_variable_declaration) =
+                    JsForVariableDeclaration::cast_ref(&possible_declaration)
+                {
+                    if js_for_variable_declaration.is_const() {
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:99]: I believe we should be able to simplify this because either:

* The parent is a list and the grand parent is the variable declaration
* Or the parent is the `JsForVariableDeclaration`

```suggestion
				if let Some(variable_declaration) = JsForVariableDeclaration::cast_ref(&possible_declaration) {
				  ... 
				} else if let Some(declaration) = JsVariableDeclaratorList::cast_ref(&possible_declaration).and_then(|declaration| declaration.syntax().parent()).and_then(JsVariableDeclaration::cast) {
					.. 
				}
```

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs (hunk 5 above)
Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:72]: Thank you @MichaReiser. I was not aware that those two types existed. I will change it accordingly.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs (hunk 6 above)
Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:96]: Because a declaration can be either JsVariableDeclaratorList or JsForVariableDeclaration, I have to do the cast for both types. Your suggestion is to remove the mutable variable, but that would require duplicating the const check.

Comment (MichaReiser) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:96]: It's correct that you have to test for both cases, but the parent only is a `JsVariableDeclaratorList` if the grandparent is a `JsVariableDeclaration`. The parent of a variable declaration in a for ... in loop is the `JsForVariableDeclaration`. 

You can see this by opening [this playground] and inspecting the syntax tree (you can click on the code on the left and it will jump to the right position in the syntax panel)

Comment (mzbac) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs]: Thanks @MichaReiser, 
Somehow I thought that the JsVariableDeclaration might be the direct parent of the JsVariableDeclarator. Thank you for explaining all of the details! 
I have updated the PR as you suggested..

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs [hunk 7]
```diff
@@ -61,13 +66,33 @@ impl Rule for NoConstAssign {
...
+
+                if let Some(js_variable_declaration) =
+                    JsVariableDeclaratorList::cast_ref(&possible_declaration)
+                        .and_then(|declaration| declaration.syntax().parent())
+                        .and_then(JsVariableDeclaration::cast)
+                {
+                    if js_variable_declaration.is_const() {
+                        return Some(declared_binding.syntax().text_trimmed_range());
+                    }
+                }
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/no_const_assign.rs:95]: Nit: You can use `else if` here because it is guaranteed that the `possible_declaration` isn't a `JsVariableDeclaratorList` if it is a `JsForVariableDeclaration`
```suggestion
                }

                else if let Some(js_variable_declaration) =
                    JsVariableDeclaratorList::cast_ref(&possible_declaration)
                        .and_then(|declaration| declaration.syntax().parent())
                        .and_then(JsVariableDeclaration::cast)
                {
                    if js_variable_declaration.is_const() {
                        return Some(declared_binding.syntax().text_trimmed_range());
                    }
                }
```

**Review by MichaReiser — APPROVED** [review]

**Comment by ematipico** [comment]
@xunilrj @leops can any of you review the PR and merge it if you see fit, please?
